### ISSUE TITLE
import/no-cycle does not detect some cycles with side-effect imports

### ISSUE THREAD

**Issue by guillaumebrunerie** [import-js/eslint-plugin-import#3147]
Here is the setup (https://codesandbox.io/p/devbox/my57j6):
```js
// a.js
import "./b";
export const f = () => globalThis.x;

// b.js
import "./c";

// c.js
import f from "./a";

globalThis.x = f();
```

This seems like a cycle to me (`a.js` expects `globalThis.x` to be set given that it imports `b.js` which imports `c.js` which sets it, but `c.js` uses an import from `a.js` in order to set `globalThis.x`).
Yet this is not detected as a cycle by `eslint-plugin-import`. Note that changing `a.js` to import `./c` instead of `./b` is detected as a cycle, so it seems to be the indirection introduced by `b.js` which confuses `no-cycle`.

**Comment by ljharb** [comment]
It’s not a cycle; it’s just a global variable (which you never want to be using).

**Comment by guillaumebrunerie** [comment]
Well, `a` imports `b` which imports `c` which imports `a`. Isn’t that a cycle? The global variable is just there to show that it creates problems.
The actual problem I’m having is Webpack complaining about "cannot access variable before initialization", which in my experience is usually due to cycles, but `import/no-cycle` doesn’t report any error.
I now managed to fix my error by shuffling things around, but it still seems to me that `import/no-cycle` should report it.

**Comment by ljharb** [comment]
Oh, fair point. Yes, that is a cycle. What's your config set to, and what version of the plugin are you using?

**Comment by guillaumebrunerie** [comment]
> Oh, fair point. Yes, that is a cycle. What's your config set to, and what version of the plugin are you using?

I’m using the latest version (2.31.0) with the latest version of ESLint (9.18.0) and the following config (taken mostly from the docs of `eslint-plugin-import`):
```js
import importPlugin from "eslint-plugin-import";
import js from "@eslint/js";

export default [
  js.configs.recommended,
  importPlugin.flatConfigs.recommended,
  {
    files: ["**/*.{js,mjs,cjs}"],
    languageOptions: {
      ecmaVersion: "latest",
      sourceType: "module",
    },
    rules: {
      "import/no-cycle": "warn",
      "import/default": "off",
    },
  },
  {
    ignores: ["dist"],
  },
];
```

**Comment by ljharb** [comment]
ok, then it should be using scc. cc @soryy708

**Comment by soryy708** [comment]
Ok we're talking about a false-negative.
Lets check if the SCC optimization is responsible for it.
Try adding [`disableScc: true`](https://github.com/import-js/eslint-plugin-import/blob/HEAD/docs/rules/no-cycle.md#disablescc). Does `no-cycle` report it as a cycle then? or still no?

**Comment by soryy708** [comment]
I have looked in the attached codesandbox.
Why does it have both an `.eslintrc.json` and an `.eslint.config.js`?
Wouldn't that confuse ESLint?

![Image](https://github.com/user-attachments/assets/ac82c301-d1ae-476d-9473-3da7a8132722)

**Comment by guillaumebrunerie** [comment]
> Ok we're talking about a false-negative. Lets check if the SCC optimization is responsible for it. Try adding [`disableScc: true`](https://github.com/import-js/eslint-plugin-import/blob/HEAD/docs/rules/no-cycle.md#disablescc). Does `no-cycle` report it as a cycle then? or still no?

It makes no difference to add `disableScc: true`, the cycle still isn’t detected.

**Comment by guillaumebrunerie** [comment]
> I have looked in the attached codesandbox. Why does it have both an `.eslintrc.json` and an `.eslint.config.js`? Wouldn't that confuse ESLint?
> 
> ![Image](https://github.com/user-attachments/assets/ac82c301-d1ae-476d-9473-3da7a8132722)

Hmm, not sure, I simply took some default codesandbox and added ESLint and `eslint-plugin-import`.
In my actual project, I do have both a `eslint.config.js` and a `.eslintrc`, but this is because the `.eslintrc` is needed for ignorePatterns to work properly with `import/no-unused-modules`.
Also, changing `a.js` to directly import `./c` instead of `./b` makes ESLint detect the cycle, so it does seem like ESLint is working properly.

**Comment by soryy708** [comment]
Cool.

In `a.js`:

```
import "./b.mjs";
```

is treated as a type-import,
we exit early in:

- https://github.com/import-js/eslint-plugin-import/blob/341178de7d0ca76e6fc35550737509b8da95d44e/src/rules/no-cycle.js#L98

**Comment by ljharb** [comment]
Why would that be a type import?

**Comment by soryy708** [comment]
I don't know. Probably shouldn't?

Perhaps because `b.js` has no exports?

**Comment by ljharb** [comment]
It definitely shouldn’t - if something with no exports is imported, it’s for side effects, and that means it’s critical and must always be considered.

**Comment by guillaumebrunerie** [comment]
> we exit early in:
> 
> https://github.com/import-js/eslint-plugin-import/blob/341178de7d0ca76e6fc35550737509b8da95d44e/src/rules/no-cycle.js#L98

Looks like this also ignores imports of the form `import { type Foo } from "foo"`, but it is worth noting that with the `verbatimModuleSyntax` TypeScript option, this form of import is used to import both a type and its side effects, so it should not be ignored.
Now I don’t know if it’s possible/desirable to have `eslint-plugin-import` behave differently depending on TypeScript options, but maybe only type imports of the form `import type { Foo } from "foo"` should be ignored by default? Or conversely there could be a `verbatimModuleSyntax` option for the `import/no-cycle` rule.

**Comment by ljharb** [comment]
We do look at tsconfig for a flag already; I'm sure we could look up verbatimModuleSyntax.

**Comment by soryy708** [comment]
So we have at least 3 false-positives here:

1. JS module style `import '...';`
2. TS `import type {...} from '...';` with `verbatimModuleSyntax: true`
3. TS `import { type ... } from '...';` with `verbatimModuleSyntax: true`

Would you like to contribute test cases?

**Comment by ljharb** [comment]
semver in eslint plugins is a bit tricky; producing more warnings can be a patch because it was a bug to miss them.

**Comment by soryy708** [comment]
apropos `verbatimModuleSyntax`:

- https://github.com/import-js/eslint-plugin-import/issues/2872
